### ISSUE TITLE
Suunto Ambit 3 Peak Data

### ISSUE THREAD

**Issue by markusdecker** [openambitproject/openambit#206]
hi.

recently the suunto ambit logs, received with openambit0.4 do not include the UTC time in the Samples anymore.

before:

            <Sample>
                <Type id="512">periodic</Type>
                <UTC>2018-11-29T18:48:46.857Z</UTC>
                <Time>33857</Time>
                <VerticalSpeed>0</VerticalSpeed>
                <Cadence>0</Cadence>
                <Temperature>202</Temperature>
                <AbsPressure>9899</AbsPressure>
                <SeaLevelPressure>10209</SeaLevelPressure>
                <Altitude>263</Altitude>
                <Distance>17</Distance>
                <Speed>65</Speed>
            </Sample>

and after 01.12.2018:

            <Sample>
                <Type id="512">periodic</Type>
                <UTC></UTC>
                <Time>53302</Time>
                <VerticalSpeed>3</VerticalSpeed>
                <Cadence>81</Cadence>
                <Temperature>186</Temperature>
                <AbsPressure>9859</AbsPressure>
                <SeaLevelPressure>10189</SeaLevelPressure>
                <Altitude>280</Altitude>
                <Distance>135</Distance>
                <Speed>267</Speed>
            </Sample>

Anyone experienced the same?

ciao 
m

**Comment by markusdecker** [comment]
hi

i wrote a small function for this specific problem and added it to the python script, just in case anyone experience the same bad thing:




```

def utcTimeFind():

    global element, time, errorc, lapDate, deltatime
    time=element.findtext("UTC")
    if time.find("T")!=-1:
     if errorc==0:
        print "utc:ok"
        errorc=1
    else:
     if errorc==0:
       print "utc:fail"
       errorc=1
     if deltatime!=None and lapDate!=None:
       tmpTime=lapDate.split("T")[1]
       tmpTime=datetime.datetime.strptime(tmpTime, "%H:%M:%S")+datetime.timedelta(milliseconds=int(deltatime))
       tmpTime=tmpTime.strftime("%H:%M:%S")
       tmpDay=lapDate.split("T")[0]

       time=tmpDay+"T"+tmpTime

    return time
```

the function calculates the time out of the delta time from the start in ms. works correct, i compared it to the gpx export from suunto. 

ciao
m

**Comment by gustavo-iniguez-goya** [comment]
um, interesting, I noticed this problem as well, but as the moves were uploaded fine, I forgot it.
However analyzing the issue, I've realized that the UTC time returned is year + 1 and the month starts at 0. QDateTime doesn't seem to like this format, so it returns an empty string.

Taking a look at the moves recorded in January 2018, the UTC date is for example: 2018-02-03T18:10:03.558Z

So I think we have to transform it (src/lib/pmem.c:1218):
```
    outtime->month = (tm->tm_mon == 0) ? 12 : tm->tm_mon - 1;
    outtime->year = (tm->tm_mon == 0) ? tm->tm_year-1 : tm->tm_year;
```
With this fix the UTC field is formatted correctly. I'll upload a move tomorrow to see if it makes any difference and post a PR if it works.

**Comment by gustavo-iniguez-goya** [comment]
well, I've uploaded a new move with this workaround applied, and I don't notice any difference. The same for the strava activity.

**Comment by markusdecker** [comment]
hi gustavo

yes, the output looks like this
```

``
Time: 2019-0-18 T 8:59
Time: 2019-0-18 T 8:59
Time: 2019-0-18 T 8:59
Time: 2019-0-18 T 8:59
Time: 2019-0-18 T 8:59
Time: 2019-0-18 T 8:59
Time: 2019-0-18 T 8:59
Time: 2019-0-18 T 8:59
Time: 2019-0-18 T 8:59
Time: 2019-0-18 T 8:59
Time: 2019-0-18 T 8:59
Time: 2019-0-18 T 8:59
Time: 2019-0-18 T 8:59
Time: 2019-0-18 T 8:59
Time: 2019-0-18 T 8:59
Time: 2019-0-18 T 8:59
Time: 2019-0-18 T 8:59
Time: 2019-0-18 T 9:0
Time: 2019-0-18 T 9:0
Time: 2019-0-18 T 9:0
Time: 2019-0-18 T 9:0
Time: 2019-0-18 T 9:0
Time: 2019-0-18 T 9:0
...
```
i hope this is helpfull.

thanks
ciao 
m

**Comment by gustavo-iniguez-goya** [comment]
Thank you Markus, that's what I expected to see. The workaround should fix it then, apply the fix and add the following lines below 
https://github.com/openambitproject/openambit/blob/master/src/libambit/pmem20.c#L1166:

```
    outtime->month = (tm->tm_mon == 0) ? 12 : tm->tm_mon - 1;
    outtime->year = (tm->tm_mon == 0) ? tm->tm_year-1 : tm->tm_year;
    printf("Time: %d-%d-%d T %d:%d\n", tm->year, tm->tm_mon, tm->tm_mday, tm->tm_hour, tm->tm_min);
    printf("OutTime: %d-%d-%d T %d:%d\n", outtime->year, outtime->month, outtime->day, outtime->hour, outtime->minute);
```

lets see what's going on there.

**Comment by markusdecker** [comment]
hi gustavo, 
you are right, it works! 
i guess i made a mistake while compiling the first version of libambit (prob used the old libambit instead?).
super. thanks.

ciao.
m

**Comment by gustavo-iniguez-goya** [comment]
great! anyway I'll try to find a better solution, because this workaround is a bit of a hack, I think.

**Comment by markusdecker** [comment]
well, its a far better solution than the python workarround. 
the next thing to solve is the interbeat interval, ....

**Comment by gustavo-iniguez-goya** [comment]
interbeat interval? open a new issue explaining it please, so we can take a look.

**Comment by rnorris** [comment]
I too was having this problem with a Suuntu Ambit 3 Multisport.

It appears to me that the time functions in pmem.c are somewhat broken and starts to go awry from December 2018 onwards.

This appears to be the proper fix:

```
diff --git a/src/libambit/pmem20.c b/src/libambit/pmem20.c
index 6e3a928..a613d50 100644
--- a/src/libambit/pmem20.c
+++ b/src/libambit/pmem20.c
@@ -1162,12 +1162,11 @@ static void add_time(ambit_date_time_t *intime, int32_t offset, ambit_date_time_
     outtime->minute = tm->tm_min;
     outtime->hour = tm->tm_hour;
     outtime->day = tm->tm_mday;
-    outtime->month = tm->tm_mon;
-    outtime->year = tm->tm_year;
+    outtime->month = tm->tm_mon + 1;
+    outtime->year = tm->tm_year + 1900;
 }

static int is_leap(unsigned int y) {
-    y += 1900;
     return (y % 4) == 0 && ((y % 100) != 0 || (y % 400) == 0);
 }
 
@@ -1180,11 +1179,11 @@ static void to_timeval(ambit_date_time_t *ambit_time, struct timeval *timeval) {
     timeval->tv_usec = 0;
     int i;
 
-    for (i = 70; i < ambit_time->year; ++i) {
+    for (i = 1970; i < ambit_time->year; ++i) {
         timeval->tv_sec += is_leap(i) ? 366 : 365;
     }
 
-    for (i = 0; i < ambit_time->month; ++i) {
+    for (i = 0; i < ambit_time->month-1; ++i) {
         timeval->tv_sec += ndays[is_leap(ambit_time->year)][i];
     }
     timeval->tv_sec += ambit_time->day - 1;
```

If I get time I'll create a proper pull request and explain the previous wrongness.

**Comment by markusdecker** [comment]
> interbeat interval? open a new issue explaining it please, so we can take a look.

@gustavo-iniguez-goya, interbeat interval is the method of the heartbeat measurement, in the suunto ambit peak 3 watch. @vascotenner did a edit of the pyhton script therefore, i dont know if this is an c-code issue?
if still needed i can open a new issue and descibe it?
what you think @vascotenner?

**Comment by vascotenner** [comment]
I think that the log files should be as close to the data exported by the watch. The python script to convert to GPX can decide how to convert the data, and what data to discard!

**Comment by centic9** [comment]
I think the underlying issue is now fixed via #212, please report possible improvement to the gpx-conversion script as separate issues if they are still useful after this fix.
